zebra-consensus/src/transaction.rs

```diff
@@ -153,175 +161,243 @@ where
+                    .await?
+                    .call(primitives::groth16::ItemWrapper::from(output).into());
+
+                async_checks.push(output_rsp.boxed());
+            }
+
+            let bvk = sapling_shielded_data.binding_verification_key();
+
+            // TODO: enable async verification and remove this block - #1939
+            {
+                let item: zebra_chain::primitives::redjubjub::batch::Item =
+                    (bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into();
+                item.verify_single().unwrap_or_else(|binding_sig_error| {
+                    let binding_sig_error = binding_sig_error.to_string();
+                    tracing::warn!(%binding_sig_error, "ignoring");
+                    metrics::counter!("zebra.error.sapling.binding",
+                                                  1,
+                                                  "kind" => binding_sig_error);
+                });
+                // Ignore errors until binding signatures are fixed
+                //.map_err(|e| BoxError::from(Box::new(e)))?;
+            }
+
+            let _rsp = redjubjub_verifier
+                .ready_and()
+                .await?
+                .call((bvk, sapling_shielded_data.binding_sig, &shielded_sighash).into())
+                .boxed();
+
+            // TODO: stop ignoring binding signature errors - #1939
+            // async_checks.push(rsp);
+        }
+
+        // Finally, wait for all asynchronous checks to complete
+        // successfully, or fail verification if they error.
+        while let Some(check) = async_checks.next().await {
+            tracing::trace!(?check, remaining = async_checks.len());
+            check?;
+        }
+
+        Ok(tx.hash())
+    }
+
+    async fn verify_v5_transaction(
+        request: &Request,
+        network: Network,
+    ) -> Result<transaction::Hash, TransactionError> {
+        Self::verify_v5_transaction_network_upgrade(request.upgrade(network))?;
+
+        // TODO:
+        // - verify transparent pool (#1981)
+        // - verify sapling shielded pool (#1981)
+        // - verify orchard shielded pool (ZIP-224) (#2105)
+        // - ZIP-216 (#1798)
+        // - ZIP-244 (#1874)
+
+        unimplemented!("V5 transaction validation is not yet complete");
+    }
+
+    fn verify_v5_transaction_network_upgrade(
+        upgrade: NetworkUpgrade,
+    ) -> Result<(), TransactionError> {
+        match upgrade {
+            // Supports V5 transactions
+            NetworkUpgrade::Nu5 => Ok(()),
+
+            // Does not support V5 transactions
+            NetworkUpgrade::Genesis
+            | NetworkUpgrade::BeforeOverwinter
+            | NetworkUpgrade::Overwinter
+            | NetworkUpgrade::Sapling
+            | NetworkUpgrade::Blossom
+            | NetworkUpgrade::Heartwood
+            | NetworkUpgrade::Canopy => {
+                Err(TransactionError::UnsupportedByNetworkUpgrade(5, upgrade))
```

Comment:
Readability: Is this the network upgrade 5 or the transaction version 5?

```rust
UnsupportedByNetworkUpgrade(5, ...)
```

Can you add a `version` method on `Transaction`, and use it during serialization as well?
(We won't want the overwintered flag in the numeric version, just the transaction version number.)

Here's an example - this V1 serialization that could be refactored:
https://github.com/ZcashFoundation/zebra/blob/5c0880810bd82855fa6d98059122112719e7de8a/zebra-chain/src/transaction/serialize.rs#L380

Reply:
I think this is great.